### DISCUSSION TITLE
Rootless Podman Without Pre-Configured subuid/subgid

### DISCUSSION THREAD

**Discussion by nlvw** [containers/podman#16018] [Ideas]
Recently the Apptainer project released a new version that no longer requires `/etc/subuid` and `/etc/subgid` to be preconfigured for running/building rootless containers.  The only requirement is that unprivileged user namespaces be enabled.  This is really convenient as AD users can both build and run containers without an admin having to setup subuid/subgid configuration.  All while being rootless with no setuid binaries.

It would be nice if rootless podman could also do this as it is unusable at scale in a SSSD (ldap/ad) environment.  Preconfiguring subuid/subgid bindings for every potential users who could login and then syncing those binding across multiple machines (10s to 100s) is not a solution.

https://github.com/apptainer/apptainer/releases/tag/v1.1.0
<img width="755" alt="image" src="https://user-images.githubusercontent.com/5723368/193375656-640ccf7e-0ae2-432a-9afa-a6946f96fb89.png">

**Comment by afbjorklund** [comment]
There is the `ignore_chown_errors` storage option, which is somewhat similar to `fakeroot` perhaps:

https://www.redhat.com/sysadmin/controlling-access-rootless-podman-users

Allowing to run without setting up subuid and subgid is a good feature, and an improvement over [docker](https://docs.docker.com/engine/security/rootless/#prerequisites)...

Otherwise there needs to be some user-accessible service, on how to "request" a mapping to be set up ?

**Reply by rhatdan** [reply]
Note this will squash all UIDs down to the single UID of the user. it also will not support changine the USER to a different UID inside of the build since this will be blocked by the OS.  IE a USER=1000 inside of a Containerfile.

**Reply by jhodrien** [reply]
fakeroot with apptainer is lovely, since it allows building of containers as a regular user without setting up subuid/subgid.  It would be very nice to be able to treat podman the same way.  Building containers from a Dockerfile with podman in this way would be hugely nice, but currently I see no way of making that work, unless anyone has any suggestions?

**Reply by rhatdan** [reply]
I would think we would need a PR to add support for fakeroot in containers-storage.
@giuseppe @flouthoc WDYT?

**Reply by giuseppe** [reply]
should that be a new mode for buildah --isolation?

Not sure you can setup an entire container using fakeroot (and we assume each RUN runs in its own container)

**Reply by rhatdan** [reply]
Sure we could try this, but fakeroot would only work if a setuid was not called during the build, correct?

**Reply by jhodrien** [reply]
Quite probably, yes.  I'm interested in whether buildah/podman can support something equivalent to the third in the list here:
https://apptainer.org/docs/user/main/fakeroot.html

**Reply by rhatdan** [reply]
Not yet, although adding something like --isolation=fakeroot, to buildah and podman build would likely be approved.

**Reply by cbs228** [reply]
> Otherwise there needs to be some user-accessible service, on how to "request" a mapping to be set up ?

We already have [oddjob-mkhomedir](https://pagure.io/oddjob/blob/master/f/src/mkhomedir.c) for creating user home directories in SSO environments. What about a mythical "`oddjob-mksubuid`" to create a /etc/subuid mapping on first login? While there are plenty of SSO environments that exceed [32k users](https://freeipa.readthedocs.io/en/latest/designs/subordinate-ids.html#design-choices), how many individual *machines* see that many different logon users? There might be some cluster head nodes that do… but how many of those also want to run rootless containers?

Assigning subuids on first-login is fraught with peril. The safest uid range to use is the one that FreeIPA already does, but that makes an eventual collision a near certainty. Since SSO machines would not share uid mappings, you couldn't just `rsync` a podman volume from one to the other. Restores from backup would also not work unless /etc/subuid was also restored. There are already problems with [rootless images and NFS](https://www.redhat.com/sysadmin/rootless-podman-nfs), and this certainly won't make them any better. The potential for accidentally cross-contaminating subuid ranges is also greater than if they were centrally-managed.

There is an experimental [pam_subuid](https://github.com/yrro/pam_subuid) already that does almost exactly what I've described. It would probably be better as a message-oriented oddjob service. That way, potential race conditions due to multiple "first" logins could be handled more gracefully.

Of course, the real "fix" would be to convince the kernel folks to widen `uid_t` to `int64_t` or larger. Then everybody on Earth could have a subuid range! (Once we manage to fix the apocalyptic impact this would have on filesystem drivers and everything else.) Yes, this is probably not going to happen.

**Reply by rhatdan** [reply]
@nalind WDYT?

**Reply by nalind** [reply]
The `--isolation` flag only factors in for builds when RUN instructions are being evaluated, so it wouldn't be exposed that way.  We'd be looking to affect how `podman` itself handles setting and verifying ownerships on layer contents, and all of that happens well before a container is even created.

My main doubt is that using LD_PRELOAD to have a shared library intercept calls into libc for a process doesn't work if the binary isn't dynamically linked, or if the shared library isn't there because the process is being run in a mount namespace where the shared library doesn't exist.  fakechroot and psuedo also use LD_PRELOAD.  fakeroot-ng uses ptrace(2), which we tend to not allow the capability for.

The API surface is large, and I doubt that any of the available options covers everything.  I would be hesitant to depend on images which were built using them.

**Reply by qhaas** [reply]
> There are already problems with [rootless images and NFS](https://www.redhat.com/sysadmin/rootless-podman-nfs), and this certainly won't make them any better. The potential for accidentally cross-contaminating subuid ranges is also greater than if they were centrally-managed.

Would a rootless podman that doesn't require subuid / subgid for pulling / creating / running containers as the same user on the host allow for `graphroot` to be set to a location on CEPH / NFS without slow FUSE based approaches (i.e. use of the overlay driver)?  I've noticed in some enterprise environment configurations that local storage with XFS / EXT4 will be small, but CEPH / NFS vast.

**Comment by eslowney** [comment]
FreeIPA/IdM already hands out subuids to users, which `getsubids $USER` correctly spits out, but podman insists on reading `/etc/subuid` for some reason.  I don't understand why it doesn't query `/etc/nsswitch.conf` and/or SSSD instead of reading files.

**Reply by rhatdan** [reply]
If this does not work that would be a serious problem.

**Reply by eslowney** [reply]
It could just be a weird local config issue.  I have a ticket open with Red Hat.  They say it should be possible.

**Reply by giuseppe** [reply]
have you restarted the podman user namespace with `podman system migrate` if you've changed the configuration for the system?

Also, please verify that podman is linked to `libsubid.so`: 

```
$ ldd $(which podman) | grep subid
	libsubid.so.4 => /lib64/libsubid.so.4 (0x00007f4c864ff000)
```

**Reply by eslowney** [reply]
Yeah, `podman system migrate` also gives the same error.  I deleted the `/etc/subuid` and `/etc/subgid` files, since it shouldn't be reading them anyways.  `ldd` does show it linked to `libsubid.so.3`.  This is on RHEL 9.3.

**Reply by eslowney** [reply]
If anyone's curious, I figured out my issue.  Even if you are using SSS in `/etc/nsswitch.conf`, the files `/etc/subuid` and `/etc/subgid` must exist, even if just empty 0-bytes files.

```
$ rm -f /etc/sub?id
$ podman system migrate
ERRO[0000] running `/usr/bin/newuidmap 12333 0 1995600002 1 1 2147483648 65536`:
Error: cannot set up namespace using "/usr/bin/newuidmap": exit status 1
$ touch /etc/subuid /etc/subgid
$ podman system migrate
$ 
```

**Reply by rhatdan** [reply]
That would seem like a bug.
@giuseppe PTAL

**Reply by giuseppe** [reply]
what do you get with the command `unshare -r --map-auto cat /proc/self/uid_map`?

**Reply by eslowney** [reply]
> what do you get with the command `unshare -r --map-auto cat /proc/self/uid_map`?

The RHEL 9.3 version of unshare (2.37.4) doesn't have `--map-auto`, but w/o it:
```
$ unshare -r cat /proc/self/uid_map
$            0 1995600002            1
```

I'm going to upgrade this box to 9.4 shortly, maybe there will be different behavior.

**Reply by giuseppe** [reply]
so newuidmap/newgidmap refuse to configure the user namespace, it doesn't depend on Podman

**Comment by septatrix** [comment]
Support for running podman without subuid/subgid files would also resolve #20040 / https://github.com/systemd/systemd/issues/29297

**Reply by giuseppe** [reply]
running podman without additional uids/gids has limitations.  You can use only one ID, so all the IDs are squashed to the same one.  You can already do that, if you set `ignore_chown_errors` in your storage.conf file.
